### PR TITLE
8318113: CSS.BackgroundImage doesn't implement equals

### PR DESCRIPTION
**Pull request by prsadhuk**
CSSBackgroundImage.equals() is implemented

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318113](https://bugs.openjdk.org/browse/JDK-8318113): CSS.BackgroundImage doesn't implement equals (**Bug** - P4)

### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Committer) ⚠️ Review applies to [d0d90aeb](https://git.openjdk.org/jdk/pull/16613/files/d0d90aebe9ab619d5de6d343a30f1c3cbb1f2206)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [d0d90aeb](https://git.openjdk.org/jdk/pull/16613/files/d0d90aebe9ab619d5de6d343a30f1c3cbb1f2206)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16613/head:pull/16613` \
`$ git checkout pull/16613`

Update a local copy of the PR: \
`$ git checkout pull/16613` \
`$ git pull https://git.openjdk.org/jdk.git pull/16613/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16613`

View PR using the GUI difftool: \
`$ git pr show -t 16613`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16613.diff">https://git.openjdk.org/jdk/pull/16613.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16613#issuecomment-1805827717)

### CODE REVIEW

**Review by aivanov-jdk — APPROVED** [review]

File: src/java.desktop/share/classes/javax/swing/text/html/CSS.java [hunk 1]
```diff
@@ -2965,6 +2965,17 @@ ImageIcon getImage(URL base) {
             }
             return image;
         }
+
+        @Override
+        public int hashCode() {
+            return (this.svalue != null) ? this.svalue.hashCode() : 0;
```
Comment (mrserb) [inline src/java.desktop/share/classes/javax/swing/text/html/CSS.java:2971]: This change look fine, but this line can be simplified by the Objects.hashCode();

**Review by TejeshR13 — APPROVED** [review]
Looks good to me.

File: test/jdk/javax/swing/text/html/CSS/CSSAttributeEqualityBug.java
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/CSS/CSSAttributeEqualityBug.java:1]: Please add 8318113 to the `@bug` tag in the test.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

**Review by prrace — APPROVED** [review]

**Review by aivanov-jdk — APPROVED** [review]
[My recent comment](https://github.com/openjdk/jdk/pull/16613#discussion_r1392808012) removed my approval.

It's *approved* after the following is addressed:

> Please add 8318113 to the @bug tag in the test.

File: src/java.desktop/share/classes/javax/swing/text/html/CSS.java (hunk 1 above)
Comment (turbanoff) [reply src/java.desktop/share/classes/javax/swing/text/html/CSS.java]: @prsadhuk idea was to replace whole expression with `Objects.hashCode`.
Current code looks confusing.

File: src/java.desktop/share/classes/javax/swing/text/html/CSS.java [hunk 2]
```diff
@@ -2965,6 +2965,17 @@ ImageIcon getImage(URL base) {
             }
             return image;
         }
+
+        @Override
+        public int hashCode() {
+            return (this.svalue != null) ? Objects.hashCode(this.svalue) : 0;
```
Comment (turbanoff) [inline src/java.desktop/share/classes/javax/swing/text/html/CSS.java:2971]:
```suggestion
            return Objects.hashCode(this.svalue);
```

**Review by turbanoff — CHANGES_REQUESTED** [review]
Use Objects.hashCode without explicit null check.

File: src/java.desktop/share/classes/javax/swing/text/html/CSS.java [hunk 3]
```diff
@@ -2968,7 +2968,7 @@ ImageIcon getImage(URL base) {
 
         @Override
         public int hashCode() {
-            return (this.svalue != null) ? Objects.hashCode(this.svalue) : 0;
+            return Objects.hashCode(this.svalue);
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/text/html/CSS.java:2971]:
```suggestion
            return Objects.hashCode(svalue);
```
Using `this` is also redundant.

File: src/java.desktop/share/classes/javax/swing/text/html/CSS.java [hunk 4]
```diff
@@ -2965,6 +2965,17 @@ ImageIcon getImage(URL base) {
...
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(this.svalue);
+        }
+
+        @Override
+        public boolean equals(Object val) {
+            return val instanceof CSS.BackgroundImage img
+                   && Objects.equals(this.svalue, img.svalue);
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/text/html/CSS.java:2977]:
```suggestion
            return val instanceof CSS.BackgroundImage img
                   && Objects.equals(svalue, img.svalue);
```
Using `this` is redundant. You didn't use it in other implementations.

**Review by aivanov-jdk — APPROVED** [review]

**Pull request closed without merging**